src/linematch.c

```diff
@@ -0,0 +1,406 @@
+#include <assert.h>
+#include <math.h>
+#include <stdbool.h>
+#include <stddef.h>
+#include <stdint.h>
+#include <string.h>
+
+#include "vim.h"           // Add Vim's main header
+
+#define LN_MAX_BUFS 8
+#define LN_DECISION_MAX 255  // pow(2, LN_MAX_BUFS(8)) - 1 = 255
+
+// struct for running the diff linematch algorithm
+typedef struct diffcmppath_S diffcmppath_T;
+struct diffcmppath_S {
+  int df_lev_score;  // to keep track of the total score of this path
+  size_t df_path_n;   // current index of this path
+  int df_choice_mem[LN_DECISION_MAX + 1];
+  int df_choice[LN_DECISION_MAX];
+  diffcmppath_T *df_decision[LN_DECISION_MAX];  // to keep track of this path traveled
+  size_t df_optimal_choice;
+};
+
+#ifdef INCLUDE_GENERATED_DECLARATIONS
+# include "linematch.c.generated.h"
+#endif
+
+static size_t line_len(const mmfile_t *m)
+{
+  char_u *s = (char_u *)m->ptr;
+  size_t n = (size_t)m->size;
+  char_u *end = strnchr(s, &n, '\n');
+  if (end) {
+    return (size_t)(end - s);
+  }
+  return (size_t)m->size;
+}
+
+#define MATCH_CHAR_MAX_LEN 800
+
+/// Same as matching_chars but ignore whitespace
+///
+/// @param s1
+/// @param s2
+static int matching_chars_iwhite(const mmfile_t *s1, const mmfile_t *s2)
+{
+  // the newly processed strings that will be compared
+  // delete the white space characters
+  mmfile_t sp[2];
+  char p[2][MATCH_CHAR_MAX_LEN];
+  for (int k = 0; k < 2; k++) {
+    const mmfile_t *s = k == 0 ? s1 : s2;
+    size_t pi = 0;
+    size_t slen = MIN(MATCH_CHAR_MAX_LEN - 1, line_len(s));
+    for (size_t i = 0; i <= slen; i++) {
+      char e = s->ptr[i];
+      if (e != ' ' && e != '\t') {
+        p[k][pi] = e;
+        pi++;
+      }
+    }
+
+    sp[k] = (mmfile_t){
+      .ptr = p[k],
```

Comment:
I don't think this type of initialization is supported by the older compilers.

Reply:
changed